### ISSUE TITLE
R2L: anaphora resolution?

### ISSUE THREAD

**Issue by williampma** [opencog/opencog#1214]
We have some form of anaphora resolution for RelEx's opencog atoms, but not for R2L atoms.  Do we not need that as well for those logic atoms?

I ask because in the process of working with SuReal, I cannot match the output of `(r2l "John steals the orange and eats the orange.")` or `(r2l "John steals the orange and eats it.")` with 

```
(EvaluationLink (stv 0.99000001 0.99000001)
	(PredicateNode "steals@d88d647d-5c02-481e-b8d0-cfd20d33fb23" (stv 0.001 0.99000001))
	(ListLink (stv 0.99000001 0.99000001)
		(ConceptNode "Bob@fd5fb72f-0d3f-4517-8b20-8147df2ee7d3" (stv 0.001 0.99000001))
		(ConceptNode "apple@ca552878-32d4-41e3-aaa3-34239ba98f34" (stv 0.001 0.99000001))
	)
)
(EvaluationLink (stv 0.99000001 0.99000001)
	(PredicateNode "eats@5ec780eb-a98b-4650-a680-3b0163a8822e" (stv 0.001 0.99000001))
	(ListLink (stv 0.99000001 0.99000001)
		(ConceptNode "Bob@fd5fb72f-0d3f-4517-8b20-8147df2ee7d3" (stv 0.001 0.99000001))
		(ConceptNode "apple@ca552878-32d4-41e3-aaa3-34239ba98f34" (stv 0.001 0.99000001))
	)
)
```
because both `(ConceptNode "apple@ca55...")` need to match to the same node, and we have two different "orange" in the first R2L output, and one "orange" and a "it" in the second.

I suppose this requires the "abstraction" post-processing pipeline at some step as well?

@AmeBel add to the Dialogue System milestone?

**Comment by anitzkin** [comment]
The "abstraction" post-processor does not work generally enough (last time I tried it) to be of much use.  I wonder if it would be better to temporarily program PLN to interpret all identically named referents as the same referent.  Then when more sophisticated anaphora resolution comes on line, you won't need to re-write all the stuff that would be using the "abstraction" technique.

**Comment by AmeBel** [comment]
@williampma  this brings back the issue u raised about ConceptNodes not being directly associated with WordNodes . I am assuming doing so would enable some type of similarity matching. Right?

**Comment by linas** [comment]
We do have anaphora resolution code within opencog. It is supposed to do this kind of abstraction. It is located in the opencog/nlp/anaphora directory.

@anitzkin when you say that the  `The "abstraction" post-processor does not work generally enough` are you talking about the anaphora resolution code, or about something else?    We really should NOT be using "something else", we really should be using the anaphora code. That is what it was designed to do...

**Comment by linas** [comment]
To be specific: when I last looked at the output of  the anaphora resolution code, it was correctly handling both of the example sentences:  `John steals the orange and eats the orange.`  and `John steals the orange and eats it.` and correctly identified both words `orange` as the same concept, for the first sentence, and the it identified `it` and `orange` for the second sentence.  There are even supposed to be unit tests that verify this, somewhere in that directory.

It had something like 2 dozen rules for performing these abstractions, and should handle a rather wide range of sentences.  Also: it was designed so that someday, we could use PLN to make far more sophisticated deductions about references.

**Comment by anitzkin** [comment]
No, Linas, I was referring to an experimental "abstraction" post-processor that William was working on that I tried to apply in my toy question-answering demo. . .

**Comment by williampma** [comment]
Yes, I am aware there is anaphora resolution code, but it is for the OpenCog RelEx output, not for R2L's output.

Perhaps linking the R2L's node to the corresponding WordInstanceNode/WordNode would help some other other process, but I am not sure it will help SuReal with the way it is envisioned.

**Comment by linas** [comment]
The anaphora resolution code was built for the relex output because you guys were not ready with R2L at that time.   Now that you are ready, it should be ported as appropriate. That was the plan all along.

Its sort of painful to hear you say what you're saying...   We had a summer-school student create this code explicitly for this project ... is there anything there that can be salvaged?  Can you explain why it doesn't fit?  What happened between last summer and now, that changed things?

**Comment by linas** [comment]
I  mean, to be clear: I envisioned this block of code as the central place where we'd be carrying out not only anaphora resolution, but also the word-sense disambiguation, and also all the markup for patient, agent, etc. It was meant to be the layer where PLN kicks in and performs reasoning.   PLN would not be running on any of the lower layers.

**Comment by williampma** [comment]
I only meant that just linking the R2L's logic nodes to the WordInstanceNode/WordNode doesn't seem to help SuReal at the current stage, not that the current anaphora resolution code is completely useless.

Now that I think more about it, I think to really know what is needed for R2L's anaphora resolution, it is best to know what the dialogue controller will output to microplanner/SuReal.  For example, if the controller want to say `Sam steals the orange` and `Sam eats the orange`, will it output something with 2 orange instances to SuReal, or will it output something with 1 orange instance.

I suspect it is the later, so maybe the way to port that to R2L is to make it so that R2L's output never use more than one instanced node for the same concept, so that `John steals the orange and eats the orange.` will have
```
(EvaluationLink
   (PredicateNode "steals@11")
   (ListLink
      (ConceptNode "John@22")
      (ConceptNode "orange@33")))
(EvaluationLink
   (PredicateNode "eats@44")
   (ListLink
       (ConceptNode "John@22")
       (ConceptNode "orange@33")))
```
then the output from the dialogue controller will match, and perhaps microplanner/SuReal will say both `Sam steals the orange` and `Sam eats the orange` in one sentence as `Same steals the orange and eats the orange` (unless the microplanner want to insert a pronoun, but in that case it also need `John steals the orange and eats it.` in the atomspace).

Humm... but then the `WordSequenceLink` uses the old `WordInstanceNode` with multiple instance of `orange`...

I guess I just want to know what's the plan for anaphora in R2L's side so that anaphora generation on microplanner/SuReal will make more sense.

**Comment by linas** [comment]
Just because a word occurs twice, you can't assume it refers to the same thing.  "John was listening to rock when he tripped on a rock": subtle syntactic clue: one is a mass noun, the other a count noun.  "John saw Fargo (the movie) when he visited Fargo, Minnesota."  Or "John was a fruit (i.e. gay) who liked fruit" -- sorry about the last one -- best I could do with orange.

"John was listening to rock and he tripped on it" implies that he really likes rock.  I dunno, there are many ambiguous cases.  There needs to be some API where you can plug in the dumbest, simplest (most likely) alternative, use that as the default, but if its wrong, then go back and try the second-ranked alternative.  So in this case, the word "orange" probably refers to the same thing, .. but maybe it doesn't, and you have to leave that possibility visible and open to higher layers to determine, just in case the simplest interpretation is wrong.  You mustn't filter out this information too early.

This was something we argued about a lot during the summer anaphora project.

**Comment by bgoertzel** [comment]
William,

If the orange that Sam steals and the orange that Sam eats are supposed to
be the same orange, then the microplanner should get one ConceptNode
covering the two references to that orange....  But of course, the two
sentences, once generated, will have two different WordInstanceNodes...

ben

On Wed, Jan 14, 2015 at 9:05 PM, William Ma <notifications@github.com>
wrote:

> I only meant that just linking the R2L's logic nodes to the
> WordInstanceNode/WordNode doesn't seem to help SuReal at the current stage,
> not that the current anaphora resolution code is completely useless.
>
> Now that I think more about it, I think to really know what is needed for
> R2L's anaphora resolution, it is best to know what the dialogue controller
> will output to microplanner/SuReal. For example, if the controller want to
> say Sam steals the orange and Sam eats the orange, will it output
> something with 2 orange instances to SuReal, or will it output something
> with 1 orange instance.
>
> I suspect it is the later, so maybe the way to port that to R2L is to make
> it so that R2L's output never use more than one instanced node for the same
> concept, so that John steals the orange and eats the orange. will have
>
> (EvaluationLink
>    (PredicateNode "steals@11")
>    (ListLink
>       (ConceptNode "John@22")
>       (ConceptNode "orange@33")))
> (EvaluationLink
>    (PredicateNode "eats@44")
>    (ListLink
>        (ConceptNode "John@22")
>        (ConceptNode "orange@33")))
>
> then the output from the dialogue controller will match, and perhaps
> microplanner/SuReal will say both Sam steals the orange and Sam eats the
> orange in one sentence as Same steals the orange and eats the orange
> (unless the microplanner want to insert a pronoun, but in that case it also
> need John steals the orange and eats it. in the atomspace).
>
> Humm... but then the WordSequenceLink uses the old WordInstanceNode with
> multiple instance of orange...
>
> I guess I just want to know what's the plan for anaphora in R2L's side so
> that anaphora generation on microplanner/SuReal will make more sense.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1214#issuecomment-70033368>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
Hummm... then it would be great to know what's the plan for anaphora resolution on R2L's output.  I had hacked around the anaphora generation problem in microplanner before, which worked with the old loose SuReal.  However, with the new SuReal solution now requiring better matches, I am a bit confused as to how the pronouns/lexical alternative should be placed into the solution.

**Comment by bgoertzel** [comment]
William --
If you have time, perhaps you could propose a candidate solution for this
issue?

On Thu, Jan 15, 2015 at 9:21 PM, William Ma <notifications@github.com>
wrote:

> Hummm... then it would be great to know what's the plan for anaphora
> resolution on R2L's output. I had hacked around the anaphora generation
> problem in microplanner before, which worked with the old loose SuReal.
> However, with the new SuReal solution now requiring better matches, I am a
> bit confused as to how the pronouns/lexical alternative should be placed
> into the solution.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1214#issuecomment-70202986>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
I haven't come up with a good solution yet.

While it would be good that R2L's output something like the following for the sentence `John steals the orange and eats the orange.` (assuming same "orange")
```
(EvaluationLink
   (PredicateNode "steals@11")
   (ListLink
      (ConceptNode "John@22")
      (ConceptNode "orange@33")))
(EvaluationLink
   (PredicateNode "eats@44")
   (ListLink
       (ConceptNode "John@22")
       (ConceptNode "orange@33")))
```
so that SuReal can match to this sentence when talking about one orange/apple/etc, it will still need some information to get to the original 2 orange's `WordInstanceNode` from the original sentence (to potential be replace by the new word orange/apple/etc at SuReal)...  kind of like
```
 (SomeSortOfLink (ConceptNode "orange@33") (WordInstanceNode "orange@33"))
 (SomeSortOfLink (ConceptNode "orange@33") (WordInstanceNode "orange@44"))
```
Then following these links SuReal can generate `Sam steals the apple and eats the apple.`


However, with the above solution, even the sentence `John steals the orange and eats it.` would produce the same R2L's output, except it would have
```
 (SomeSortOfLink (ConceptNode "orange@33") (WordInstanceNode "orange@33"))
 (SomeSortOfLink (ConceptNode "orange@33") (WordInstanceNode "it@44"))
```
Suppose SuReal want to replace "orange" with "apple", then the resulting sentence would end up as `Sam steals the apple and eats apple`, which is incorrect because `apple` doesn't really match to `it` (in terms of grammar), but SuReal doesn't know by looking at the R2L's output graph.


So maybe the solution should be to keep the R2L's output as is, with 2 different orange instance for the same concept
```
(EvaluationLink
   (PredicateNode "steals@11")
   (ListLink
      (ConceptNode "John@22")
      (ConceptNode "orange@33")))
(EvaluationLink
   (PredicateNode "eats@44")
   (ListLink
       (ConceptNode "John@22")
       (ConceptNode "orange@44")))
```

but how would SuReal match to this graph when it only want to talk about one `apple` concept...  it is like making one variable maps to two nodes in one solution, which sounds kind of strange... so at the moment I am leaning toward somehow making the first solution work.

Perhaps SuReal need to generate those `SomeSortOfLink` as part of its query... but there's also the sentence `John steals and eats the orange`, which really only does have one orange instance, WordInstanceNode included.

And then the problem is reversed when generating anaphora.  I will have to think some more.
